### ISSUE TITLE
Lora caused memory leak for simple finetuning 

### ISSUE THREAD

**Issue by Randolph-zeng** [huggingface/peft#832]
### System Info

Issues persisted across several peft version 0.3, 0.4, 0.5(dev)
Accelerator I used are 0.21.0,
Pytorch version I tried are 1.13.0 and 2.0 and both experience same memory explosion 
transformers == 4.31.0

### Who can help?

@pacman100 , @younesbelkada  @sayakpaul 

### Information

- [X] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder
- [X] My own task or dataset (give details below)

### Reproduction

To reproduce, all you need is the trl huggingface library, llama2 weight and the LIMA datasets(from huggingface datasets)

```
import torch, sys
from trl import SFTTrainer
from peft import LoraConfig
from datasets import load_dataset
from transformers import LlamaForCausalLM, LlamaTokenizer, TrainingArguments, create_optimizer


def formatting_func(examples):
    output_texts = []
    for i in range(len(examples['conversations'])):
        text = f"### Question: {examples['conversations'][i][0]}\n ### Answer: {examples['conversations'][i][1]}"
        output_texts.append(text)
    return output_texts

def main():
    model_path = '/home/zzs/llama2/Llama-2-7b-hf'
    tokenizer_path = '/home/zzs/llama2/Llama-2-7b-hf'
    dataset = load_dataset("GAIR/lima")
    single_turn_train_dataset = dataset['train'].filter(lambda obj: len(obj['conversations'])==2)

    tokenizer = LlamaTokenizer.from_pretrained(tokenizer_path)
    tokenizer.pad_token = tokenizer.eos_token
    # leave the detailed-control to accelerate config 
    model = LlamaForCausalLM.from_pretrained(model_path)
    batch_size, num_epochs = 8, 3
    # Some args are from LIMA paper setup
    peft_config = LoraConfig(
        r=16,
        lora_alpha=32,
        lora_dropout=0.05,
        bias="none",
        task_type="CAUSAL_LM",
    )
    
    train_args = TrainingArguments(
        output_dir='/data/llama_ckpt',
        adam_beta1=0.9,
        adam_beta2=0.95,
        weight_decay=0.1,
        learning_rate=1e-5,
        logging_steps=10,
        save_strategy='epoch',
        save_total_limit=3,
        per_device_train_batch_size=int(batch_size/8),
        num_train_epochs=num_epochs,
        fp16=True,
        fp16_full_eval=True
    )

    trainer = SFTTrainer(
        model,
        tokenizer=tokenizer,
        args=train_args,
        packing=False,
        train_dataset=single_turn_train_dataset,
        formatting_func=formatting_func,
        # peft_config=peft_config,
        max_seq_length=2048,
    )



    trainer.train()

if __name__ == "__main__":
    main()
```

### Expected behavior

I am trying to do a 7b llama finetune, and it should take around 15GB to load a half precision model to memory and since I am using lora on q,v proj, the extra memory to do finetune should not be too large if I am only using batch_size = 1 and max_seq_len 512. However, the memory consumption can goes to 60 GB.
I managed to pinpoint the issues and it seems to be caused by lora.Linear module. Everytime there is a subsequent matmul operations on the output of q_proj, or v_proj, the memory consumptions goes up for a few hunderd megabyets. I understand Pytorch need to keep some computation graphs for the intermediate computation but the amount of memory increase is just not explainable by normal tracking of gradient computation.
Any insight will be greatly appreciated!

**Comment by BenjaminBossan** [comment]
This sounds like a very severe issue, thank you for reporting.

Just to be sure that I understood correctly: When the training begins, the memory use is in line with expectations, but it increases steadily during training, (presumably) caused by a memory leak in the LoRA linear layer. Is that right?

> the memory consumption can goes to 60 GB

Does it plateau at 60 GB or will it increase even higher but you stopped there?

Also, in the code you pasted, `peft_config` is commented out, but I assume this was just for testing and that the memory leak only occurs with peft?

I tried reproducing the issue using more simple architectures (MLP, BERT), trying both fp32 and fp16, and there was no memory leak, total consumption was totally flat. So whatever is causing the leak is not a general issue, but has to do with either the model (llama2 in this case), the hyper-parameters, the data, or the trainer.

**Comment by Randolph-zeng** [comment]
@BenjaminBossan Hi Benjamin thanks for the speedy response and reproducing ! Here is what I found after some further dig in:
My colleague told me that he can successfully finetune the same model with the same peft/accelerate/transformers libraries. I tried his code and found that the issue is caused by the loading method.  With the load_in_8bit flag on and off, the memory  consumption for a 7b model can range from the 13GB to 51GB with all the other variables staying the same.
So I suspect it is a problem more on the transformers side ? I will dig in more to see why this happens. I will close the issue for now. Thanks a lot for the help !

![image](https://github.com/huggingface/peft/assets/11933185/53986afb-b001-41bb-b419-cf2ed307f3d5)
![image](https://github.com/huggingface/peft/assets/11933185/9f64f69c-c390-4aea-be7e-34ea85aed2db)

**Comment by BenjaminBossan** [comment]
Thanks for investigating further.

> With the load_in_8bit flag on and off, the memory consumption for a 7b model can range from the 13GB to 51GB with all the other variables staying the same.
> So I suspect it is a problem more on the transformers side ?

Is this finding true regardless of whether PEFT is used or not?

**Comment by Randolph-zeng** [comment]
Yes, on both case, the PEFT was on, I am planning to raise an issue on transformers library. If you are interested, this is the script I am testing and this line is the only thing I changed from the script https://github.com/tloen/alpaca-lora/blob/main/finetune.py#L114 :)

**Comment by treeaaa** [comment]
First, here's my guess: the sudden surge in memory might be partly caused by the following section:
https://github.com/huggingface/peft/blob/7d99466446d65219a401a2c64f1d84d1b1be31b4/src/peft/utils/other.py#L97-L101

My environment setup is as follows:
(GCP) A100-40G(main) & A100-80G
+ (continer) [nvidia/cuda   11.8.0-cudnn8-devel-ubuntu22.04](https://hub.docker.com/layers/nvidia/cuda/11.8.0-cudnn8-devel-ubuntu22.04/images/sha256-ee127a83b5269251476a3a1933972735a0a8a3269d35fda374f548212687d5db?context=explore)
+ python3.10
+ (hg)llama-2 7G
+ [text-generation-webui](https://github.com/oobabooga/text-generation-webui)

(using the text-generation-webui on the A100-40G)
I first observed a GPU memory consumption of 13,740MiB when loading the model by using the command
```bash
watch -n 1 nvidia-smi 
```

Then, during the LoRA training,I observed 13,740MiB -> 33,306Mib -> CUDA out of memory
I find it incredibly baffling. Does this mean that LoRA consumed 20G of my GPU memory?

To monitor the GPU memory usage, I integrated the following code snippet at various points in the code:
```python
logger.info(f"{torch.cuda.memory_reserved()}")
```

Upon tracing, I found that the spike was caused by this block:
https://github.com/huggingface/peft/blob/7d99466446d65219a401a2c64f1d84d1b1be31b4/src/peft/utils/other.py#L97-L101
Initially, I loaded the model in float16 which roughly used 7G2 = 14G of GPU memory. This code converts the float16 model to float32, leading to a model size of 7G4 = 28G. Thus, the 33,306MiB observed via nvidia-smi likely comprises the ~28G model size plus PyTorch's reservation.

Furthermore, after inserting code.at line 102, I observed a usage of 26,781MiB, which is consistent with my expectations.
```python
torch.cuda.empyt_cache()
```

Even after emptying the cache, I still encountered an OOM error. Eventually, I found this memory calculation tutorial:
https://github.com/huggingface/transformers/issues/25572#issuecomment-1687749561

I then switched to an A100-80G VM, and I was finally able to fine-tune, observing a memory usage of 44,431MiB, which is close to the estimate from the aforementioned article.


Conclusion:
1.I suspect the sudden increase in memory usage is related to prepare_model_for_kbit_training in peft.
2.I couldn't replicate the 50G memory usage, suggesting there might be other underlying issues.

**Comment by waylongo** [comment]
I also have the same problem. This is caused by `prepare_model_for_kbit_training` func that converts fp16 to fp32. I guess this function is designed for quantized (INT8 or INT4) pretrained models .
